src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -7640,7 +7640,7 @@ public <T> T unwrap(Class<T> iface) throws SQLException {
     private volatile SQLWarning originalSqlWarnings;
 
     /** open statements */
-    private List<ISQLServerStatement> openStatements;
+    private volatile List<ISQLServerStatement> openStatements;
```

Comment:
@divang Here is the reason we are making them volatile

bIsClosed and openStatements are not volatile, and as a result a thread creating statements while under a request ( started using beginRequest) may add them in openStatements list, with some other thread closing them but not removing from openStatements just due to that thread having read a stale value(null) of openStatements( and other combinations of issues of this sort). Hence, we are making bIsClosed and openStatements volatile.

Reply:
As discussed offline, incorporating resources from the Java Concurrency in Practice book by Doug Lea and Brian Goetz will enhance understanding. 

Synchronized blocks and methods can ensure that operations execute atomically, but it is a common misconception that synchronized is only about atomicity or demarcating "critical sections". Synchronization also has another significant and subtle aspect: "memory visibility". [chapter:3.1 Sharing Objects]

Intrinsic locking can be used to guarantee that one thread sees the effects of another in a predictable manner: [chapter 3.1.3: Locking and Visibility]
![image](https://github.com/user-attachments/assets/dcc6654b-6506-4d94-9704-7b45c7cbad2c)

**Locking is not just about mutual exclusion**, it is **also about memory visibilit**y. To ensure that **all threads see the most up-to-date values of shared mutable variables**, the reading and writing threads must synchronize on a common lock. [chapter:3.1.3 Sharing Objects]

About Volatile Variables, accessing a volatile variable performs "on locking" and so cannot cause the executing thread to block, making volatile variable a **"lighter weight synchronization mechanism**" than synchronized. [chapter:3.1.4 Volatile Variables]

ReentrantLock implements Lock, providing the same mutual exclusion and memory-visibility guarantees as synchronized. [Chapter 13.1 Lock and ReentrantLock]

As we discussed, we may simulate above scenarios and validate those concepts.

Reply:
@divang 

openStatement's usage is always being guarded under the lock, hence we do not need it to be volatile. 

requestStarted and bIsClosed booleans are being used without locks, they will be needed to be volatile to avoid the potential invalid state of statements being already closed but not part of openStatements list.